### PR TITLE
Reducing list API payload

### PR DESCRIPTION
**Pull request by anushka567**
### Description
Fetching only selected attributes using [query.SetAttrSelection](https://github.com/googleapis/google-cloud-go/blob/storage/v1.45.0/storage/storage.go#L1784) for ListObjects API which are sufficient for serving the request and reduce the payload for network call.
Also included :  necessary changes in the fake bucket tests.

Perf Metrics for listing : 
| Test Directory Size ( in number of files) | Mean Latency(Master) (in msec) | Mean Latency(reduced payload) (in msec) | Percentage Improvement |
| ------------------- | ---------------------- | --------------------------------- | ------------------------- |  
| 1000 | 61.153  | 56.325| 7.89% |
| 10000 | 462.775 | 410.701 | 11.25%|
| 50000 |  2373.293 |2103.235 | 11.38%|
| 100000 | 4677.908 | 4354.92| 6.90% |
| 200000 |9417.683 |8628.266 | 8.38% |

Note: More detailed metrics [here](https://docs.google.com/spreadsheets/d/101rbvNWjU55sTUaNP9p86D_g1A_2vxCjq7FSmyPDuzo/edit?usp=sharing)
### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/storage/bucket_handle.go [hunk 1]
```diff
@@ -300,6 +300,12 @@ func (bh *bucketHandle) ListObjects(ctx context.Context, req *gcs.ListObjectsReq
 		IncludeFoldersAsPrefixes: req.IncludeFoldersAsPrefixes,
 		//MaxResults: , (Field not present in storage.Query of Go Storage Library but present in ListObjectsQuery in Jacobsa code.)
 	}
+	err = query.SetAttrSelection([]string{"Name", "Size", "Generation", "Metageneration", "Updated", "Metadata"})
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle.go:303]: We need to fetch all fields which are in gcs.MinObject . contentEncoding and crc32 are missing.

File: internal/storage/fake/testing/bucket_tests.go [hunk 2]
```diff
@@ -543,15 +543,35 @@ func (t *createTest) ObjectAttributes_Default() {
...
+			ForceFetchFromGcs:              true,
+			ReturnExtendedObjectAttributes: true})
 	AssertEq(nil, err)
 
-	AssertThat(listing.CollapsedRuns, ElementsAre())
-	AssertEq("", listing.ContinuationToken)
-
-	AssertEq(1, len(listing.Objects))
-	ExpectThat(listing.Objects[0], DeepEquals(o))
+	ExpectThat(minObj.Name, Equals(o.Name))
```
Comment (vadlakondaswetha) [inline internal/storage/fake/testing/bucket_tests.go:554]: can you extract this into private methods - comparing minObject and extendedObjectAttributes

File: internal/storage/bucket_handle.go (hunk 1 above)
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle.go:303]: what about unit tests in bucket_handle_test.go?

Comment (anushka567) [reply internal/storage/bucket_handle.go]: Unit tests in bucket_handle_test.go only assert on object names and do not reference attributes other than these attributes which are fetched. So , the tests do not fail and no additional tests is required.

Comment (anushka567) [reply internal/storage/bucket_handle.go]: Added.

File: internal/storage/fake/testing/bucket_tests.go (hunk 2 above)
Comment (anushka567) [reply internal/storage/fake/testing/bucket_tests.go]: Updated.

File: internal/storage/fake/testing/bucket_tests.go [hunk 3]
```diff
@@ -425,6 +425,31 @@ func (t *bucketTest) matchesStartTime(start time.Time) Matcher {
 	return timeutil.TimeNear(start, slop)
 }
 
+// Given gcs.minObject and gcs.extendedObjectAttributes , assert on attributes of gcs.Object
+func (t *bucketTest) assertOnObjectAttributes(minObj *gcs.MinObject, extendedAttr *gcs.ExtendedObjectAttributes, o *gcs.Object) {
```
Comment (vadlakondaswetha) [inline internal/storage/fake/testing/bucket_tests.go:429]: rename this to expectedMinObj and expectedExtendedAttr

File: internal/storage/bucket_handle.go [hunk 4]
```diff
@@ -334,6 +334,12 @@ func (bh *bucketHandle) ListObjects(ctx context.Context, req *gcs.ListObjectsReq
 		IncludeFoldersAsPrefixes: req.IncludeFoldersAsPrefixes,
 		//MaxResults: , (Field not present in storage.Query of Go Storage Library but present in ListObjectsQuery in Jacobsa code.)
 	}
+	err = query.SetAttrSelection([]string{"Name", "Size", "Generation", "Metageneration", "Updated", "Metadata", "ContentEncoding", "CRC32C"})
```
Comment (vadlakondaswetha) [inline internal/storage/bucket_handle.go:337]: I see this method is returning gcs.Object but gcs.MinObject is inserted into cache.
Do we want to change the callers of this method to use gcs.MinObject directly? You can do that in a followup cl.

File: internal/storage/bucket_handle.go (hunk 1 above)
Comment (vadlakondaswetha) [reply internal/storage/bucket_handle.go]: If fake-gcs-server is handling setAttrSelection field, then please add testcases for checking if other properties are not returned.

File: internal/storage/fake/testing/bucket_tests.go (hunk 3 above)
Comment (anushka567) [reply internal/storage/fake/testing/bucket_tests.go]: Done.

File: internal/storage/bucket_handle.go (hunk 1 above)
Comment (anushka567) [reply internal/storage/bucket_handle.go]: fake-gcs-server is not handling setAttrSelection field.It is returning all attributes when listing .

File: internal/storage/bucket_handle.go (hunk 4 above)
Comment (anushka567) [reply internal/storage/bucket_handle.go:337]: Doing this in this [PR](https://github.com/GoogleCloudPlatform/gcsfuse/pull/2657)

**Review by vadlakondaswetha — APPROVED** [review]

**Pull request merged**